### PR TITLE
[BIOMAGE-1864] added heartbeat to tasks

### PR DESCRIPTION
**Pull request by kafkasl**
# Description
Added parallel processing to set up a heartbeat function that sends heartbeats to the state machines during the task processing.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1846

#### Link to staging deployment URL (or set N/A)
https://ui-yes-heartbeat.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/api/pull/333

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [x] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [x] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: pipeline-runner/init.R [hunk 1]
```diff
@@ -350,6 +362,10 @@ init <- function() {
         flog.appender(appender.tee(file.path(dump_folder, "logs.txt")))
 
         tryCatchLog({
+
+                plan(multiprocess)
```
Comment (gerbeldo) [inline pipeline-runner/init.R:366]:
```suggestion
                plan(list(multiprocess, sequential))
```

File: pipeline-runner/init.R [hunk 2]
```diff
@@ -282,6 +283,43 @@ call_data_processing <- function(task_name, input, pipeline_config) {
...
 }
 
+#
+# start_heartbeat(task_token, aws_config)
+# IN task_token, aws_config
+#
+# Sends a heartbeat to the state machine every 10 seconds
+# Once the task is completed the heartbeat will fail accordingly with a
+# task timeout and exit the loop and a new heartbeat will be set up by next task.
+# This method is invoked with `r_bg` which creates a new process which does not inherit
```
Comment (gerbeldo) [inline pipeline-runner/init.R:293]:
```suggestion
# This method is invoked with `callr::r_bg` which creates a new process which does not inherit
```

File: pipeline-runner/init.R [hunk 3]
```diff
@@ -282,6 +283,43 @@ call_data_processing <- function(task_name, input, pipeline_config) {
     return(message_id)
 }
 
+#
+# start_heartbeat(task_token, aws_config)
+# IN task_token, aws_config
+#
+# Sends a heartbeat to the state machine every 10 seconds
```
Comment (aerlaut) [inline pipeline-runner/init.R:290]: 30 seconds

File: pipeline-runner/init.R [hunk 4]
```diff
@@ -282,6 +283,43 @@ call_data_processing <- function(task_name, input, pipeline_config) {
...
+# Once the task is completed the heartbeat will fail accordingly with a
+# task timeout and exit the loop and a new heartbeat will be set up by next task.
```
Comment (aerlaut) [inline pipeline-runner/init.R:292]: The task timeout is 10800 seconds (3 hours). Does this mean that this process will run for 3 hours before failing?

File: pipeline-runner/init.R (hunk 3 above)
Comment (aerlaut) [inline pipeline-runner/init.R:290]: 30 seconds

File: pipeline-runner/init.R [hunk 5]
```diff
@@ -350,11 +389,23 @@ init <- function() {
...
 
         tryCatchLog({
+
+                # start heartbeat as a different process in the background
+                message("Starting heartbeat")
+                # message inside r_bg will ONLY be printed into /tmp/[out|err]
+                # to see them
+                # 1. log into the R container
+                # 2. cat /tmp/out  o tail -f /tmp/out
+                r_bg(func=start_heartbeat, args=list(
```
Comment (aerlaut) [inline pipeline-runner/init.R:399]: `r_bg` actually returns a [process object](https://callr.r-lib.org/reference/r_bg.html#value) which has a [kill function ](https://www.tidyverse.org/blog/2018/09/processx-3.2.0/). Maybe you can use this function to kill the heartbeat process at the end of task completion?

File: pipeline-runner/init.R (hunk 4 above)
Comment (kafkasl) [reply pipeline-runner/init.R:292]: no:
1. as soon as the task is completed the heartbeat fails and the loop stops
2. when a pipeline finishes the pod and everything running inside is just killed

File: pipeline-runner/init.R (hunk 5 above)
Comment (kafkasl) [reply pipeline-runner/init.R:399]: sounds good will do

File: pipeline-runner/init.R [hunk 6]
```diff
@@ -349,12 +390,26 @@ init <- function() {
         dump_folder <- file.path(DEBUG_PATH, debug_prefix)
         flog.appender(appender.tee(file.path(dump_folder, "logs.txt")))
 
+        # start heartbeat as a different process in the background
+        message("Starting heartbeat")
+        # message inside r_bg will ONLY be printed into /tmp/[out|err]
+        # to see them
+        # 1. log into the R container
+        # 2. cat /tmp/out  o tail -f /tmp/out
```
Comment (aerlaut) [inline pipeline-runner/init.R:398]:
```suggestion
        # 2. cat /tmp/out or tail -f /tmp/out
```

**Review by aerlaut — APPROVED** [review]
Minor change to comment

**Pull request merged**